Pirania UI

luandro opened the pull request:
Adds user interface for Pirania based on the designs from [Figma](https://www.figma.com/file/auIzuiYMoZcFicqtd0MvU1/LimeApp---General?node-id=0%3A1).

The consumed API expecs the [Pirania new API](https://github.com/libremesh/lime-packages/pull/893) ubus API.

### Known TODOs:

- [x] add node information to vouchers
- [x] add ability to disable voucher
- [ ] add translations

hiurequeiroz committed:
feat: pirania ui

luandro committed:
fix(routes): fixes CommunityProtectedRoute to pass props down

luandro committed:
feat(pirania): adds interface for pirania based on new-pirania-api

luandro commented on plugins/lime-plugin-pirania/src/piraniaApi.js:
```diff
@@ -0,0 +1,59 @@
+	return api
+		.call("pirania", "list_vouchers", {})
+		.then((response) => response.vouchers.map((voucher) => {
+			const { is_active, expiration_date } = voucher;
+			let status = is_active ? "used" : "available";
+			if (expiration_date < Date.now()) status = "disabled";
```
This isn't right. Confused about the right way to check the status. Tips @spiccinini @germanferrero ?

luandro committed:
chore(pirania): removed uneeded tests for voucher filter

luandro committed:
chore(pirania): adds disabling vouchers and better status logic

luandro committed:
chore(pirania): adds author_node information with default sorting

luandro commented on plugins/lime-plugin-pirania/src/piraniaApi.js:
```diff
@@ -0,0 +1,64 @@
+	return api
+		.call("pirania", "list_vouchers", {})
+		.then((response) =>
+			response.vouchers.map((voucher) => {
+				const { is_active, expiration_date, duration_m } = voucher;
+				let status = is_active ? "used" : "available";
+
+				if ((expiration_date && expiration_date < now) || duration_m === 0)
+					status = "disabled";
+				return {
+					...voucher,
+					status
+				};
+			})
+		)
```
This is how I'm checking the status of the voucher. Is this the best way? Related to: https://github.com/libremesh/lime-packages/pull/893#discussion_r738762672

spiccinini replied:
In https://github.com/libremesh/lime-packages/pull/893/commits/36784d36b468577cdc96f0052cad0b429c04a72f I've added the `status` attribute that may be one of 'available', 'active' or 'expired'

luandro commented on src/components/app.js:
```diff
@@ -45,26 +45,22 @@ const Routes = () => (
 			.filter(plugin => plugin.additionalRoutes)
 			.map(plugin => plugin.additionalRoutes)
 			.flat()
-			.map(([path, Component]) => {
```
Having lots of linting problems, `npx eslint --fix src` doesn't seem to be working correctly. Any tips @germanferrero?

germanferrero replied:
@luandro I've not paid to much attention to linting lately. Maybe the way to go is to properly configure https://prettier.io/ for the project, and do a massive linting PR. Do you have any experience with it?

luandro replied:
I've got some experience with setting up Prettier + Husky for pre-commit linting. I could create another branch/PR for that.

germanferrero replied:
That would be awesome!!!

germanferrero commented on plugins/lime-plugin-pirania/pirania.spec.js:
```diff
@@ -0,0 +1,145 @@
+	it("let you filter by all vouchers", async () => {
+		const select = await screen.findByLabelText("filter by");
+		expect(select).toHaveValue("all-vouchers");
+		expect(await (await screen.findAllByText("used")).length).toBeGreaterThan(
```
This double await statement shouldn't be needed. The same applies for the following similar cases.

germanferrero commented on plugins/lime-plugin-pirania/pirania.spec.js:
```diff
@@ -0,0 +1,145 @@
+		).toBeGreaterThan(0);
+	});
+	it("let you filter by used vouchers", async () => {
+		const select = await screen.findByLabelText("filter by");
```
For this test to be complete we could (maybe should) test that <VoucherListItem /> is rendered only for those vouchers matching the given filter. For that we could make use of data-test-id  with a composed value like "voucher-list-item-\<code\>" and query them with findtByTestId.

germanferrero commented on plugins/lime-plugin-pirania/pirania.spec.js:
```diff
@@ -0,0 +1,145 @@
+});
+
+describe("voucher list search field", () => {
+	beforeEach(() => {
```
Here we could also use the same approach proposed for the filter to tests that only those vouchers matching the search criteria are rendered.

germanferrero commented on plugins/lime-plugin-pirania/src/components/voucherListItem.spec.js:
```diff
@@ -0,0 +1,79 @@
+		).toBeInTheDocument();
+	});
+
+	it("renders voucher-list-item with a test code", async () => {
```
I think that this tests title is a left over from the idea of using data-test-id for testing filtering and searching in the vouchers list screen

germanferrero commented on plugins/lime-plugin-pirania/src/components/voucherListItem.spec.js:
```diff
@@ -0,0 +1,79 @@
+
+	it("routes to the voucher detail screen on click", async () => {
+		render(<VoucherListItem {...voucher} />);
+		const element = await screen.findByRole("voucher-item");
```
Not sure if we are adding a role attribute to the component thinking of [ARIA roles for assistive technologies](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles) or we where only looking for a way to test that if you click in the voucher element then the routing happens.
In the latter case, we could use getByTestId instead, or even getByTextId (testing click in the voucher code for instance).

germanferrero commented on plugins/lime-plugin-pirania/src/components/voucherListItem.spec.js:
```diff
@@ -0,0 +1,79 @@
+				status={"used"}
+			/>
+		);
+		expect(await screen.findByText("used")).toBeInTheDocument();
```
"active" is the wording that we have chosen for vouchers that are being used, isn't it?
If we want to change it, I wouldn't vote for "used", maybe "in use".. but instead  "active" sounds good to me.

germanferrero commented on plugins/lime-plugin-pirania/src/components/voucherListItem.spec.js:
```diff
@@ -0,0 +1,79 @@
+		render(<VoucherListItem {...voucher} status={"disabled"} />);
+		expect(await screen.findByText("disabled")).toBeInTheDocument();
+	});
+	it("shows expired date when it has an expiration date", async () => {
```
It would be nice that the test describes how the text for expiration should be built. Like "Expires at: <formated_date>" or "Expires: <formated_date>"

germanferrero commented on plugins/lime-plugin-pirania/src/components/voucherListItem.spec.js:
```diff
@@ -0,0 +1,79 @@
+	permanent: false
+};
+
+describe("voucher list item", () => {
```
For the case of "availabe" / "non active yet" vouchers there is an activation_deadline property which let users create vouchers that need to be activated before that deadline for activation to succeed.
I think we could add this date in the UI, similarly as we show the expiration date for active vouchers.

germanferrero commented on plugins/lime-plugin-pirania/src/screens/voucher.spec.js:
```diff
@@ -0,0 +1,78 @@
+	}
+];
+
+vouchers[0].status = vouchers[0].is_active
```
The logic of how the voucher status is defined shouldn't be coded in the test file.

luandro replied:
Status coming from backend now.

germanferrero commented on plugins/lime-plugin-pirania/src/components/copy.js:
```diff
@@ -0,0 +1,40 @@
+					height="15"
+					style={{ cursor: "pointer", opacity: isCopied ? 0.3 : 1 }}
+					src={
+						"data:image/svg+xml;base64,PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLTgiIHN0YW5kYWxvbmU9Im5vIj8+CjxzdmcKICAgZGF0YS1uYW1lPSJMYXllciAxIgogICB2aWV3Qm94PSIwIDAgNTcgNjAiCiAgIHg9IjBweCIKICAgeT0iMHB4IgogICB2ZXJzaW9uPSIxLjEiCiAgIGlkPSJzdmc5MjIiCiAgIHNvZGlwb2RpOmRvY25hbWU9Im5vdW5fY29weV80MDM5NDQ1LnN2ZyIKICAgd2lkdGg9IjU3IgogICBoZWlnaHQ9IjYwIgogICBpbmtzY2FwZTp2ZXJzaW9uPSIxLjEuMSAoM2JmNWFlMGQyNSwgMj ...
```
I'd prefer to reduce bundle size, (and sacrifice a bit of beauty) by using a unicode character for copy-paste symbol like this one:
https://unicode-table.com/en/29C9/
More options are available at:
https://stackoverflow.com/questions/8714969/is-there-a-unicode-character-for-copy-and-paste

germanferrero commented on plugins/lime-plugin-pirania/pirania.spec.js:
```diff
@@ -0,0 +1,145 @@
+
+describe("voucher list", () => {
+	it("shows a button to create a new voucher that routes to creation page", async () => {
+		render(<VoucherList vouchers={vouchers} />);
```
Let please keep the pattern of testing components that render entire screen, having no properties passed by beside url parameters, and mocking backend apis.
That let us keep track of how things connect to each other.

In this case, we could test PiraniaPage instead of VoucherList and mock getVouchersList api for each case.

germanferrero reviewed (changes requested):
Good woooooork! :tada: @luandro and the team! I'm really happy that you are contributing again to the LimeApp :slightly_smiling_face:

I did a general review, focusing on tests files, and found that we need to clarify a bit more the different status in which a voucher may be in it's lifecycle. We may address that first, and then I can come back to review the rest of the code in more detail.

So AFAIU, the possible status are:

"available":  
The voucher has not been activated by anyone yet. In this case it may have an activation deadline. It would be nice to show it as 'Activated', 'Activation available until <date>'
"active":
The voucher is already associated with a device mac.
It may be a permanent voucher, or have an expiration date.
"expired":
The voucher has been activated and then it reach it's `duration`.
Or the voucher wasn't activated before its activation deadline.
Or the voucher has been manualy invalidated.

I also think that this status should be computed in the backend and included
in the api response as `status`: ['availabe', 'active', 'expired']

spiccinini commented:
> status



> So AFAIU, the possible status are:
> 
> "available": The voucher has not been activated by anyone yet. In this case it may have an activation deadline. It would be nice to show it as 'Activated', 'Activation available until ' "active": The voucher is already associated with a device mac. It may be a permanent voucher, or have an expiration date. "expired": The voucher has been activated and then it reach it's `duration`. Or the voucher wasn't activated before its activation deadline. Or the voucher has been manualy invalidated.
> 
> I also think that this status should be computed in the backend and included in the api response as `status`: ['availabe', 'active', 'expired']

Yes I think this is how I understand it and I also agree to add this status in the backend. @luandro is this ok ?

luandro committed:
chore(pirania): changes base64 icons to unicode

germanferrero committed:
create-plugin: fix api example generator

germanferrero committed:
remoteSupport spec: use render from utils

germanferrero committed:
improvement(pirania): add pirania activation ui

luandro committed:
chore(pirania): adds loading when creating voucher

luandro committed:
fix(pirania): correctly shows create button on voucher list page

germanferrero commented on plugins/lime-plugin-pirania/src/components/createVoucherForm.js:
```diff
@@ -2,9 +2,11 @@ import { h } from "preact";
 import { useState } from "preact/hooks";
 import I18n from "i18n-js";
 import style from "../style.less";
+import Loading from "../../../../src/components/loading";
```
We set up an alias for `components`, so we can simply import like:
`import Loading from "components/loading";

I researched for 5' if there is an automation/linting tool for ensuring imports use aliases but I didn't find anything convincing.

germanferrero commented on plugins/lime-plugin-pirania/src/components/createVoucherForm.js:
```diff
@@ -2,9 +2,11 @@ import { h } from "preact";
 const maxLength = 100;
 
 const CreateVoucherForm = ({ submitVoucher }) => {
+	const [loading, setLoading] = useState(false);
```
ReactQuery's mutations give us handy variables to track the query status like isLoading. Which you can use instead of setting up a useState hook by your own.

Example:
```
const [addVoucher, {isLoading}] = useAddVoucher()
````

germanferrero commented on plugins/lime-plugin-pirania/src/components/createVoucherForm.js:
```diff
@@ -20,8 +22,12 @@ const CreateVoucherForm = ({ submitVoucher }) => {
+	};
 	return (
-		<form class={style.createForm} onSubmit={(e) => submitVoucher(e, input)}>
+		<form class={style.createForm} onSubmit={(e) => createVoucher(e)}>
```
We are using [react-hook-form](https://react-hook-form.com/) to build forms, handle validation, etc.

There are some screens in the project already using it, did you come across any of them? Do you think it would simplify the code for this form?

germanferrero committed:
chore(pirania): remove voucher activation config

germanferrero committed:
feat(portal editor): add portal editor

germanferrero committed:
improvement(pirania): add img compression

germanferrero committed:
improvement(pirania): loading state for slow compression

germanferrero committed:
improvement(pirania): style portal editor

germanferrero committed:
improvement(pirania): add field validation

germanferrero committed:
chore(pirania): fix link to portal editor

germanferrero committed:
chore(pirania): fix config page url

germanferrero committed:
chore(pirania): fix set_portal_config api

germanferrero committed:
chore(pirania): request url starts with http[s]

germanferrero committed:
chore(pirania): add color, fix logo default

germanferrero committed:
chore(pirania): add to plugins

germanferrero committed:
chore(pirania): adapt to lingui

germanferrero committed:
chore(pirania): adapt list-create vouchers to lingui

germanferrero committed:
chore(pirania): timeago add locales

germanferrero committed:
chore(pirania): voucherListItem handle backend status

germanferrero committed:
chore(pirania): more styling for voucherItem

germanferrero committed:
chore(pirania): update test with author_name

germanferrero committed:
chore(pirania): move author to description row

germanferrero committed:
chore(pirania): more styling for vouchers list

germanferrero committed:
chore(pirania): complete tests&code for vouchers lists

Filters vouchers created in this node by default.
Tests that voucher item for each filtered voucher is shown.

germanferrero committed:
chore(pirania): add tests for search

germanferrero committed:
chore(pirania): make copy work on http

germanferrero committed:
chore(pirania): set input type for search input

germanferrero committed:
chore(pirania): complete test&code for voucher creation

germanferrero committed:
chore(pirania): fix body for create voucher

germanferrero committed:
chore(pirania): sort list by recently created

germanferrero committed:
chore(pirania): fix deadline utc offset

Takes timezone offset into account

germanferrero committed:
chore(pirania): move deadline to 23:59

germanferrero committed:
chore(pirania): improve post create screen

germanferrero committed:
chore(pirania): improve detail screen

germanferrero committed:
chore(pirania): separate edit and invalidate in two screens.

In this way we can alert users that invalidation cannot be undone

spiccinini committed:
refactor(pirania): update to pirania-app package merged into pirania

germanferrero committed:
chore(pirania): support for invalidated vouchers

germanferrero committed:
chore(pirania): force reload of portal config after set

germanferrero committed:
pirania: fix confusing type conversion

germanferrero committed:
chore(pirania): fix dateToLocalUnixTimestamp offset

germanferrero committed:
chore(pirania): unify texts, less translations

germanferrero committed:
chore(translations): extract pirania translations

Also complete pirania spanish translations

germanferrero committed:
chore(pirania): adding back voucher usage config

germanferrero committed:
chore(pirania): send with_voucher config

germanferrero committed:
chore(pirania): update portal content after set
